File: julia/2022/day_11.jl
mutable struct Monkey
    items::Vector{UInt}
    op::Char
    op_val::Union{UInt, Nothing}
    divisor::UInt
    true_id::UInt
    false_id::UInt
    num_inspections::UInt
end

function parse_monkey(str::AbstractString)::Monkey
    lines = split(str, '\n')
    m = match(r"new = old ([+*]) (\d+|old)", lines[3])
    op = only(m[1])
    if m[2] == "old"
        op_val = nothing
    else
        op_val = parse(UInt, m[2])
    end
    Monkey(
        [parse(UInt, s.match) for s in eachmatch(r"\d+", lines[2])],
        op,
        op_val,
        (parse(UInt, match(r"\d+", s).match) for s in lines[4:6])...,
        UInt(0)
    )
end

function do_round(monkeys::Vector{Monkey}; advanced::Bool)
    # We can mod by the LCM of all divisors without affecting throw decisions
    cap = lcm((m.divisor for m in monkeys)...)
    for monkey in monkeys
        monkey.num_inspections += length(monkey.items)
        for item in monkey.items
            b = something(monkey.op_val, item)
            item = monkey.op == '+' ? item + b : item * b
            if advanced
                item %= cap
            else
                item ÷= 3
            end
            dest = item % monkey.divisor == 0 ? monkey.true_id : monkey.false_id
            # Need to add 1 since Julia indices start at 1
            push!(monkeys[dest+1].items, item)
        end
        empty!(monkey.items)
    end
end

function business(monkeys::Vector{Monkey}; advanced::Bool)::UInt
    num_rounds = advanced ? 10_000 : 20
    monkeys = deepcopy(monkeys)
    for _ in 1:num_rounds
        do_round(monkeys, advanced=advanced)
    end
    partialsort!(monkeys, 2, by=(m -> m.num_inspections), rev=true)
    prod(m.num_inspections for m in monkeys[1:2])
end

function main()
    text = read("input/2022/input_11.txt", String)
    monkeys = [parse_monkey(s) for s in eachsplit(text, "\n\n", keepempty=false)]
    for (p, advanced) in enumerate([false, true])
        println("Part $p: $(business(monkeys, advanced=advanced))")
    end
end

main()

File: kotlin/y2022/Day11.kt
package y2022

import java.io.File

data class Monkey(
    val items: MutableList<ULong>,
    val inspect: (ULong) -> ULong,
    val divisor: ULong,
    val trueId: Int,
    val falseId: Int,
    var numInspected: ULong = 0U,
) {
    fun clone(): Monkey = this.copy(items = this.items.toMutableList())
}

fun findInt(s: String): Int = Regex("""\d+""").find(s)!!.value.toInt()

fun parseMonkey(s: String): Monkey {
    val lines = s.lines()
    val match = Regex("""new = old ([+*]) (\d+|old)""").find(lines[2])!!
    val (operator, operand) = match.destructured
    val binaryOp: (ULong, ULong) -> ULong = when (operator) {
        "+" -> ULong::plus
        "*" -> ULong::times
        else -> throw UnsupportedOperationException()
    }
    val unaryOp: (ULong) -> ULong = when (val opVal = operand.toULongOrNull()) {
        null -> { x -> binaryOp(x, x) }
        else -> { x -> binaryOp(x, opVal) }
    }
    return Monkey(
        Regex("""\d+""").findAll(lines[1]).map { it.value.toULong() }.toMutableList(),
        unaryOp,
        findInt(lines[3]).toULong(),
        findInt(lines[4]),
        findInt(lines[5]),
    )
}

fun business(monkeys: MutableList<Monkey>, advanced: Boolean): ULong {
    val numRounds = if (advanced) { 10_000 } else { 20 }
    val manageFn: (ULong) -> ULong = if (advanced) {
        val cap = monkeys.map { it.divisor }.reduce(ULong::times);
        { it % cap }
    } else {
        { it / 3U }
    }
    repeat(numRounds) {
        for (monkey in monkeys) {
            monkey.numInspected += monkey.items.size.toULong()
            for (item in monkey.items) {
                val newItem = manageFn(monkey.inspect(item))
                val dest = when (newItem % monkey.divisor) {
                    0UL -> monkey.trueId
                    else -> monkey.falseId
                }
                monkeys[dest].items.add(newItem)
            }
            monkey.items.clear()
        }
    }
    monkeys.sortByDescending { it.numInspected }
    return monkeys[0].numInspected * monkeys[1].numInspected
}

fun main() {
    val text = File("input/2022/input_11.txt").readText().trim()
    val monkeys = text.split("\n\n").map(::parseMonkey)
    for (part in 1..2) {
        val advanced = part == 2
        println("Part $part: ${business(monkeys.map(Monkey::clone).toMutableList(), advanced)}")
    }
}
